internal/uvm/network.go

```diff
@@ -1,31 +1,129 @@
 )
 
-// AddNetNS adds network namespace inside the guest & adds endpoints to the guest on that namepace
-func (uvm *UtilityVM) AddNetNS(id string, endpoints []*hns.HNSEndpoint) (err error) {
+var (
+	// ErrNetNSAlreadyAttached is an error indicating the guest UVM already has
+	// an endpoint by this id.
+	ErrNetNSAlreadyAttached = errors.New("network namespace already added")
+	ErrNetNSNotFound        = errors.New("network namespace not found")
+)
+
+// AddNetNS adds network namespace inside the guest.
+//
+// If a namespace with `id` already exists returns `ErrNetNSAlreadyAttached`.
+func (uvm *UtilityVM) AddNetNS(id string) error {
+	logrus.WithFields(logrus.Fields{
+		logfields.UVMID: uvm.id,
+		"netns-id":      id,
+	}).Debug("uvm::AddNetNS")
+
+	uvm.m.Lock()
+	defer uvm.m.Unlock()
+	if _, ok := uvm.namespaces[id]; ok {
+		return ErrNetNSAlreadyAttached
+	}
+
+	if uvm.isNetworkNamespaceSupported() {
+		// Add a Guest Network namespace. On LCOW we add the adapters
+		// dynamically.
+		if uvm.operatingSystem == "windows" {
+			hcnNamespace, err := hcn.GetNamespaceByID(id)
+			if err != nil {
+				return err
+			}
+			guestNamespace := hcsschema.ModifySettingRequest{
+				GuestRequest: guestrequest.GuestRequest{
+					ResourceType: guestrequest.ResourceTypeNetworkNamespace,
+					RequestType:  requesttype.Add,
+					Settings:     hcnNamespace,
+				},
+			}
+			if err := uvm.Modify(&guestNamespace); err != nil {
+				return err
+			}
+		}
+	}
+
+	if uvm.namespaces == nil {
+		uvm.namespaces = make(map[string]*namespaceInfo)
+	}
+	uvm.namespaces[id] = &namespaceInfo{
+		nics: make(map[string]*nicInfo),
+	}
+	return nil
+}
+
+// AddEndpointsToNS adds all unique `endpoints` to the network namespace
+// matching `id`. On failure does not roll back any previously successfully
+// added endpoints.
+//
+// If no network namespace matches `id` returns `ErrNetNSNotFound`.
+func (uvm *UtilityVM) AddEndpointsToNS(id string, endpoints []*hns.HNSEndpoint) error {
+	logrus.WithFields(logrus.Fields{
+		logfields.UVMID: uvm.id,
+		"netns-id":      id,
+	}).Debug("uvm::AddEndpointsToNS")
+
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
-	ns := uvm.namespaces[id]
-	if ns == nil {
-		ns = &namespaceInfo{}
 
+	ns, ok := uvm.namespaces[id]
+	if !ok {
+		return ErrNetNSNotFound
+	}
+
+	for _, endpoint := range endpoints {
+		if _, ok := ns.nics[endpoint.Id]; !ok {
+			nicID := guid.New()
+			if err := uvm.addNIC(nicID, endpoint); err != nil {
+				return err
+			}
+			ns.nics[endpoint.Id] = &nicInfo{
+				ID:       nicID,
+				Endpoint: endpoint,
+			}
+		}
+	}
+	return nil
+}
+
+// RemoveNetNS removes the namespace from the uvm and all remaining endpoints in
+// the namespace.
+//
+// If a namespace matching `id` is not found this command silently succeeds.
+func (uvm *UtilityVM) RemoveNetNS(id string) error {
+	logrus.WithFields(logrus.Fields{
+		logfields.UVMID: uvm.id,
+		"netns-id":      id,
+	}).Debug("uvm::RemoveNetNS")
+
+	uvm.m.Lock()
+	defer uvm.m.Unlock()
```

Comment:
And again here about length of holding lock. Probably not so important in the tear-down path though.

Reply:
Same